arangod/Replication2/StateMachines/Document/DocumentFollowerState.cpp

```diff
@@ -100,82 +98,39 @@ auto DocumentFollowerState::applyEntries(
-              auto const& collectionId = doc.collectionId;
-              self->_transactionHandler->abortTransactionsForShard(doc.shardId);
-              auto res = data.core->dropShard(doc.shardId, collectionId);
-              if (res.fail()) {
-                LOG_CTX("8c7a0", FATAL, self->loggerContext)
-                    << "Failed to drop shard " << doc.shardId
-                    << " of collection " << collectionId
-                    << " with error: " << res;
-                FATAL_ERROR_EXIT();
-              }
-              LOG_CTX("cdc44", TRACE, self->loggerContext)
-                  << "Dropped local shard " << doc.shardId << " of collection "
-                  << collectionId;
-            } else {
-              // Even though a shard was dropped before acquiring the snapshot,
-              // we could still see transactions referring to that shard.
-              if (!data.core->isShardAvailable(doc.shardId)) {
-                LOG_CTX("1970d", INFO, self->loggerContext)
-                    << "Will not apply transaction " << doc.tid << " for shard "
-                    << doc.shardId << " because it is not available";
-                continue;
-              }
-
-              auto res = self->_transactionHandler->applyEntry(doc);
-              if (res.fail()) {
-                LOG_CTX("1b08f", FATAL, self->loggerContext)
-                    << "Failed to apply entry " << entry->first
-                    << " to local shard " << doc.shardId
-                    << " with error: " << res;
-                FATAL_ERROR_EXIT();
-              }
-
-              if (doc.operation == OperationType::kAbortAllOngoingTrx) {
-                self->_activeTransactions.clear();
-                releaseIndex =
-                    self->_activeTransactions.getReleaseIndex(entry->first);
-              } else if (doc.operation == OperationType::kCommit ||
-                         doc.operation == OperationType::kAbort) {
-                self->_activeTransactions.erase(doc.tid);
-                releaseIndex =
-                    self->_activeTransactions.getReleaseIndex(entry->first);
-              } else {
-                self->_activeTransactions.emplace(doc.tid, entry->first);
-              }
+            auto [index, doc] = *entry;
+
+            auto currentReleaseIndex = std::visit(
+                [&data, index = entry->first](
+                    auto&& op) -> ResultT<std::optional<LogIndex>> {
+                  return data.applyEntry(op, index);
+                },
+                doc.getInnerOperation());
+
+            if (currentReleaseIndex.fail()) {
+              LOG_CTX("0aa2e", FATAL, self->loggerContext)
+                  << "failed to apply entry " << doc
+                  << " on follower: " << currentReleaseIndex.result();
+              FATAL_ERROR_EXIT();
+            }
+            if (currentReleaseIndex->has_value()) {
+              releaseIndex = std::move(currentReleaseIndex).get();
```

Comment:
Looks like `releaseIndex` is only set here, so we don't really need it and can return `currentReleaseIndex` instead.

Reply:
I use `releaseIndex` in order to preserve only the last `currentReleaseIndex` that's calculated in the while loop, I can't just return `currentReleaseIndex` directly.

Reply:
Ah, good point. 🤦‍♂️